subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/DynamicVersionResolver.java

```diff
@@ -224,7 +224,11 @@ void resolve() {
                     resolvedVersionMetadata.failed(versionListingResult.result.getFailure());
                     break;
                 case Listed:
-                    selectMatchingVersionAndResolve();
+                    try {
+                        selectMatchingVersionAndResolve();
+                    } catch (Throwable t) {
+                        failed(new ModuleVersionResolveException(dependency.getSelector(), t));
```

Comment:
In which case `Throwable` would be thrown here?

Reply:
See the comment on this commit: 81af24cf39fa5e1d0aed70684bab24af41854a73
This call involve execution of user-provided rules, as well as HTTP resource access, so we don't know what failure might result. Previously we had a [try-catch block in `DynamicVersionResolver`](https://github.com/gradle/gradle/pull/3382/files#diff-d9c9dda104b95d531eaa79ebc4dd2ba4R136) but I removed that.

Reply:
I don't think this should be done. Instead I think we should keep the previous `try/catch` block several lines above. If we catch here, what if this catches a "critical failure"? Should we process it in the same way as above? This looks a small change, but I'm not sure the potential influence. For a patch release, I think we should make as little change as possible.

Reply:
Yeah, it's a good point that we should minimize this change.
The reason I added this here is because `IvyDynamicRevisionRemoteResolveWithMetadataSupplierIntegrationTest` was failing without it. (The failure message had changed when the Metadata Supplier rule threw an exception). I was trying to make the loop in `DynamicVersionResolver` more similar to the [one in `RepositoryChainComponentMetadataResolver`](https://github.com/gradle/gradle/blob/81af24cf39fa5e1d0aed70684bab24af41854a73/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java#L135-L163). (note no try/catch around line 138).

Anyway, I'll put the original catch block back, and remove the second commit from this PR.

Reply:
That would be good. After this I think this PR is good to go.